# San Diego County Kindergarten Vaccination Rates Analysis
by [Joe Yerardi](https://twitter.com/JoeYerardi)  

*inewsource* conducted an analysis of data from the [California Department of Education](http://www.cde.ca.gov/ds/sd/sd/filessp.asp) and the [California Department of Public Health](http://www.shotsforschool.org/k-12/reporting-data/) to determine vaccination rates at kindergartens in San Diego County for the 2016-17 school year.

The analysis found that the proportion of San Diego County kindergartners fully up-to-date on their vaccinations reached its highest point in more than a decade and that the gaps between wealthier and poorer schools vanished in the wake of a 2015 law severely curtailing the ability of parents to acquire personal belief exemptions for their children.

These results formed the basis for the April 28, 2017 story "[Kindergartner vaccinations rising in San Diego County after state law change](https://wp.me/p6tuYl-6kJ)."

In [1]:
import geopy
import pandas as pd
import requests
# Suppress scientific notation and set the precision of float values to two
pd.set_option('display.float_format', lambda x: '%.2f' % x)
# Display all columns in dataframes
pd.set_option('display.max_columns', None)
# Display all rows in dataframes
pd.set_option('display.max_rows', None)

## Import kindergarten vaccinations data

We've got some data to import, starting with the Excel spreadsheet of 2016-17 kindergarten vaccinations data.

In [2]:
# Read the school code column as a string, specify the columns to import and assign new column names
vax1617 = pd.read_excel('Data/vaccinations1617.xlsx', converters={'SCHOOL CODE':str}, parse_cols = 'A:V',
                        names=('school_code','county','pub_priv','district','city','school','enrollment',
                               'uptodate','uptodate_pct','uptodate_suppressed','conditional','conditional_pct',
                               'conditional_suppressed','pme','pme_pct','pme_suppressed','pbe','pbe_pct',
                               'pbe_suppressed','others','others_pct','others_suppressed'))
vax1617.info() # 6,469 records, as expected

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6469 entries, 0 to 6468
Data columns (total 22 columns):
school_code               6469 non-null object
county                    6469 non-null object
pub_priv                  6469 non-null object
district                  5581 non-null object
city                      6469 non-null object
school                    6469 non-null object
enrollment                6469 non-null int64
uptodate                  3183 non-null float64
uptodate_pct              3183 non-null float64
uptodate_suppressed       3286 non-null object
conditional               3183 non-null float64
conditional_pct           3183 non-null float64
conditional_suppressed    3286 non-null object
pme                       3183 non-null float64
pme_pct                   3183 non-null float64
pme_suppressed            3286 non-null object
pbe                       3183 non-null float64
pbe_pct                   3183 non-null float64
pbe_suppressed            3286 non-null 

In [3]:
vax1617.head(1)

school_code          county pub_priv district      city        school  \
0     6975080  SAN BERNARDINO  PRIVATE      NaN  HIGHLAND  ST. ADELAIDE   

   enrollment  uptodate  uptodate_pct uptodate_suppressed  conditional  \
0          49      9.00         18.37                 NaN         0.00   

   conditional_pct conditional_suppressed  pme  pme_pct pme_suppressed  pbe  \
0             0.00                    NaN 0.00     0.00            NaN 0.00   

   pbe_pct pbe_suppressed  others  others_pct others_suppressed  
0     0.00            NaN    0.00        0.00               NaN

We want to do some year-to-year comparisons between schools so let's also import the 2015-16 kindergarten vaccinations data.

In [4]:
# Read the school code column as a string, specify the columns to import and assign new column names
vax1516 = pd.read_excel('data/vaccinations1516.xlsx', converters={'SCHOOL CODE':str}, parse_cols = 'A:O',
                        names=('school_code','county','pub_priv','district','city','school','enrollment','uptodate',
                        'uptodate_pct','conditional','conditional_pct','pme','pme_pct','pbe','pbe_pct'))
vax1516.info() # 7,422 records, as expected

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7422 entries, 0 to 7421
Data columns (total 15 columns):
school_code        7422 non-null object
county             7422 non-null object
pub_priv           7422 non-null object
district           5784 non-null object
city               7422 non-null object
school             7422 non-null object
enrollment         7137 non-null float64
uptodate           7137 non-null float64
uptodate_pct       7137 non-null float64
conditional        7137 non-null float64
conditional_pct    7137 non-null float64
pme                7137 non-null float64
pme_pct            7137 non-null float64
pbe                7137 non-null float64
pbe_pct            7137 non-null float64
dtypes: float64(9), object(6)
memory usage: 869.8+ KB


In [5]:
vax1516.head(1)

school_code   county pub_priv                            district     city  \
0     0130419  ALAMEDA   PUBLIC  ALAMEDA COUNTY OFFICE OF EDUCATION  HAYWARD   

                     school  enrollment  uptodate  uptodate_pct  conditional  \
0  ALAMEDA COUNTY COMMUNITY         nan       nan           nan          nan   

   conditional_pct  pme  pme_pct  pbe  pbe_pct  
0              nan  nan      nan  nan      nan

## Import school lunch data

Now let's import the Excel spreadsheet of free and reduced price lunch figures.

In [6]:
# Read the school code column as a string, specify the sheet and columns to import, skip the first row after the headers and assign new column names
lunch = pd.read_excel('data/lunch1617.xlsx', sheetname='FRPM School-Level Data ', skiprows=1, converters={'School Code':str},
                      parse_cols = 'D,G,V', names=('school_code','school','frpm_pct'))
lunch.info() # 10,478 records, as expected

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10478 entries, 0 to 10477
Data columns (total 3 columns):
school_code    10478 non-null object
school         10478 non-null object
frpm_pct       10478 non-null float64
dtypes: float64(1), object(2)
memory usage: 245.7+ KB


In [7]:
lunch.head(1)

school_code                                  school  frpm_pct
0     0112607  Envision Academy for Arts & Technology      0.76

## Merge, filter and analyze the data

Let's start by merging the two years of vaccinations data.

In [8]:
vax = pd.merge(vax1617, vax1516, on='school_code', how='left', suffixes=('_1617', '_1516'))
vax.head(1)

school_code     county_1617 pub_priv_1617 district_1617 city_1617  \
0     6975080  SAN BERNARDINO       PRIVATE           NaN  HIGHLAND   

    school_1617  enrollment_1617  uptodate_1617  uptodate_pct_1617  \
0  ST. ADELAIDE               49           9.00              18.37   

  uptodate_suppressed  conditional_1617  conditional_pct_1617  \
0                 NaN              0.00                  0.00   

  conditional_suppressed  pme_1617  pme_pct_1617 pme_suppressed  pbe_1617  \
0                    NaN      0.00          0.00            NaN      0.00   

   pbe_pct_1617 pbe_suppressed  others  others_pct others_suppressed  \
0          0.00            NaN    0.00        0.00               NaN   

  county_1516 pub_priv_1516 district_1516 city_1516 school_1516  \
0         NaN           NaN           NaN       NaN         NaN   

   enrollment_1516  uptodate_1516  uptodate_pct_1516  conditional_1516  \
0              nan            nan                nan               nan   

   conditional_pct_1516  pme_1516  pme_pct_1516  pbe_1516  pbe_pct_1516  
0                   nan       nan           nan       nan           nan

Now let's filter the two years of kindergarten vaccinations data down to just those schools in San Diego County.

In [9]:
vax_sd = vax[vax['county_1617'] == 'SAN DIEGO']
vax_sd.head(1)

school_code county_1617 pub_priv_1617 district_1617  city_1617  \
2     7080427   SAN DIEGO       PRIVATE           NaN  SAN DIEGO   

                   school_1617  enrollment_1617  uptodate_1617  \
2  ISLAMIC SCHOOL OF SAN DIEGO               40          20.00   

   uptodate_pct_1617 uptodate_suppressed  conditional_1617  \
2              50.00                 NaN              0.00   

   conditional_pct_1617 conditional_suppressed  pme_1617  pme_pct_1617  \
2                  0.00                    NaN      0.00          0.00   

  pme_suppressed  pbe_1617  pbe_pct_1617 pbe_suppressed  others  others_pct  \
2            NaN      0.00          0.00            NaN    0.00        0.00   

  others_suppressed county_1516 pub_priv_1516 district_1516  city_1516  \
2               NaN   SAN DIEGO       PRIVATE           NaN  SAN DIEGO   

                   school_1516  enrollment_1516  uptodate_1516  \
2  ISLAMIC SCHOOL OF SAN DIEGO            24.00          24.00   

   uptodate_pct_1516  conditional_1516  conditional_pct_1516  pme_1516  \
2             100.00              0.00                  0.00      0.00   

   pme_pct_1516  pbe_1516  pbe_pct_1516  
2          0.00      0.00          0.00

And now merge the vaccinations data with the school lunch data.

In [10]:
vax_sd = pd.merge(vax_sd, lunch, on='school_code', how='left', suffixes=('_vax', '_lunch'))
vax_sd.info() # 509 records returned, as expected

<class 'pandas.core.frame.DataFrame'>
Int64Index: 509 entries, 0 to 508
Data columns (total 38 columns):
school_code               509 non-null object
county_1617               509 non-null object
pub_priv_1617             509 non-null object
district_1617             450 non-null object
city_1617                 509 non-null object
school_1617               509 non-null object
enrollment_1617           509 non-null int64
uptodate_1617             269 non-null float64
uptodate_pct_1617         269 non-null float64
uptodate_suppressed       240 non-null object
conditional_1617          269 non-null float64
conditional_pct_1617      269 non-null float64
conditional_suppressed    240 non-null object
pme_1617                  269 non-null float64
pme_pct_1617              269 non-null float64
pme_suppressed            240 non-null object
pbe_1617                  269 non-null float64
pbe_pct_1617              269 non-null float64
pbe_suppressed            240 non-null object
others        

Finally, it's time to analyze.  
  
Our first question: Which schools saw the greatest increase in the proportion of their kindergarteners that are up-to-date on their vaccinations?

In [11]:
vax_sd['pct_change_uptodate'] = ((vax_sd['uptodate_pct_1617'] - vax_sd['uptodate_pct_1516']) / vax_sd['uptodate_pct_1516']) * 100
vax_sd.sort_values('pct_change_uptodate', ascending=False).head()

school_code county_1617 pub_priv_1617            district_1617  city_1617  \
148     0127118   SAN DIEGO        PUBLIC        DEHESA ELEMENTARY     ALPINE   
79      0108548   SAN DIEGO        PUBLIC        SAN DIEGO UNIFIED  SAN DIEGO   
224     6996045   SAN DIEGO       PRIVATE                      NaN  SAN DIEGO   
229     7102098   SAN DIEGO       PRIVATE                      NaN   CARLSBAD   
239     6118723   SAN DIEGO        PUBLIC  ALPINE UNION ELEMENTARY     ALPINE   

                         school_1617  enrollment_1617  uptodate_1617  \
148              THE HEIGHTS CHARTER               35          33.00   
79                     IFTIN CHARTER               66          58.00   
224          MARIA MONTESSORI SCHOOL               21          17.00   
229        SANDERLING WALDORF SCHOOL               21           9.00   
239  CREEKSIDE EARLY LEARNING CENTER              244         229.00   

     uptodate_pct_1617 uptodate_suppressed  conditional_1617  \
148              94.29                 NaN              0.00   
79               87.88                 NaN              8.00   
224              80.95                 NaN              2.00   
229              42.86                 NaN              1.00   
239              93.85                 NaN              9.00   

     conditional_pct_1617 conditional_suppressed  pme_1617  pme_pct_1617  \
148                  0.00                    NaN      2.00          5.71   
79                  12.12                    NaN      0.00          0.00   
224                  9.52                    NaN      0.00          0.00   
229                  4.76                    NaN      5.00         23.81   
239                  3.69                    NaN      1.00          0.41   

    pme_suppressed  pbe_1617  pbe_pct_1617 pbe_suppressed  others  others_pct  \
148            NaN      0.00          0.00            NaN    0.00        0.00   
79             NaN      0.00          0.00            NaN    0.00        0.00   
224            NaN      2.00          9.52            NaN    0.00        0.00   
229            NaN      6.00         28.57            NaN    0.00        0.00   
239            NaN      3.00          1.23            NaN    2.00        0.82   

    others_suppressed county_1516 pub_priv_1516            district_1516  \
148               NaN   SAN DIEGO        PUBLIC        DEHESA ELEMENTARY   
79                NaN   SAN DIEGO        PUBLIC        SAN DIEGO UNIFIED   
224               NaN   SAN DIEGO       PRIVATE                      NaN   
229               NaN   SAN DIEGO       PRIVATE                      NaN   
239               NaN   SAN DIEGO        PUBLIC  ALPINE UNION ELEMENTARY   

     city_1516                      school_1516  enrollment_1516  \
148     ALPINE              THE HEIGHTS CHARTER            24.00   
79   SAN DIEGO                    IFTIN CHARTER            46.00   
224  SAN DIEGO          MARIA MONTESSORI SCHOOL            34.00   
229   CARLSBAD        SANDERLING WALDORF SCHOOL            32.00   
239     ALPINE  CREEKSIDE EARLY LEARNING CENTER           214.00   

     uptodate_1516  uptodate_pct_1516  conditional_1516  conditional_pct_1516  \
148           3.00              12.50             19.00                 79.17   
79           21.00              45.65             24.00                 52.17   
224          17.00              50.00             15.00                 44.12   
229           9.00              28.13              0.00                  0.00   
239         137.00              64.02             76.00                 35.51   

     pme_1516  pme_pct_1516  pbe_1516  pbe_pct_1516  \
148      0.00          0.00      2.00          8.33   
79       0.00          0.00      1.00          2.17   
224      0.00          0.00      2.00          5.88   
229      0.00          0.00     23.00         71.88   
239      0.00          0.00      1.00          0.47   

                              school  frpm_pct  pct_change_uptodate  
1

It looks like The Heights Charter in the Dehesa Elementary school district saw the greatest overall increase in the proportion of their kindergarteners that are up-to-date on their vaccinations. How about just the personal belief exemptions?

In [12]:
vax_sd['pct_change_pbe'] = ((vax_sd['pbe_pct_1617'] - vax_sd['pbe_pct_1516']) / vax_sd['pbe_pct_1516']) * 100
vax_sd.sort_values(['pct_change_pbe', 'pbe_pct_1516'], ascending=[True, False]).head()

school_code county_1617 pub_priv_1617      district_1617        city_1617  \
153     6203491   SAN DIEGO       PRIVATE                NaN  RANCHO SANTA FE   
266     0106799   SAN DIEGO        PUBLIC  SAN DIEGO UNIFIED          LA MESA   
145     6927883   SAN DIEGO       PRIVATE                NaN     SOLANA BEACH   
159     0118083   SAN DIEGO        PUBLIC  SAN DIEGO UNIFIED        SAN DIEGO   
155     6119168   SAN DIEGO        PUBLIC  SAN DIEGO UNIFIED        SAN DIEGO   

                       school_1617  enrollment_1617  uptodate_1617  \
153                   HORIZON PREP               31          25.00   
266        LEARNING CHOICE ACADEMY               54          29.00   
145     SANTA FE CHRISTIAN SCHOOLS               38          35.00   
159            INNOVATIONS ACADEMY               46          35.00   
155  SAN DIEGO COOPERATIVE CHARTER               48          42.00   

     uptodate_pct_1617 uptodate_suppressed  conditional_1617  \
153              80.65                 NaN              3.00   
266              53.70                 NaN              0.00   
145              92.11                 NaN              1.00   
159              76.09                 NaN              4.00   
155              87.50                 NaN              0.00   

     conditional_pct_1617 conditional_suppressed  pme_1617  pme_pct_1617  \
153                  9.68                    NaN      3.00          9.68   
266                  0.00                    NaN      0.00          0.00   
145                  2.63                    NaN      2.00          5.26   
159                  8.70                    NaN      7.00         15.22   
155                  0.00                    NaN      6.00         12.50   

    pme_suppressed  pbe_1617  pbe_pct_1617 pbe_suppressed  others  others_pct  \
153            NaN      0.00          0.00            NaN    0.00        0.00   
266            NaN      0.00          0.00            NaN   25.00       46.30   
145            NaN      0.00          0.00            NaN    0.00        0.00   
159            NaN      0.00          0.00            NaN    0.00        0.00   
155            NaN      0.00          0.00            NaN    0.00        0.00   

    others_suppressed county_1516 pub_priv_1516      district_1516  \
153               NaN   SAN DIEGO       PRIVATE                NaN   
266               NaN   SAN DIEGO        PUBLIC  SAN DIEGO UNIFIED   
145               NaN   SAN DIEGO       PRIVATE                NaN   
159               NaN   SAN DIEGO        PUBLIC  SAN DIEGO UNIFIED   
155               NaN   SAN DIEGO        PUBLIC  SAN DIEGO UNIFIED   

           city_1516                    school_1516  enrollment_1516  \
153  RANCHO SANTA FE                   HORIZON PREP            40.00   
266        SAN DIEGO        LEARNING CHOICE ACADEMY            56.00   
145     SOLANA BEACH     SANTA FE CHRISTIAN SCHOOLS            42.00   
159        SAN DIEGO            INNOVATIONS ACADEMY            40.00   
155        SAN DIEGO  SAN DIEGO COOPERATIVE CHARTER            49.00   

     uptodate_1516  uptodate_pct_1516  conditional_1516  conditional_pct_1516  \
153          27.00              67.50              0.00                  0.00   
266          33.00              58.93              6.00                 10.71   
145          29.00              69.05              1.00                  2.38   
159          26.00              65.00              3.00                  7.50   
155          36.00              73.47              3.00                  6.12   

     pme_1516  pme_pct_1516  pbe_1516  pbe_pct_1516  \
153      0.00          0.00     13.00         32.50   
266      0.00          0.00     17.00         30.36   
145      0.00          0.00     12.00         28.57   
159      0.00          0.00     11.00         27.50   
155      0.00          0.00     10.00         20.41   

                            school  frpm_pct  pct_change_uptodate  \
153                       

Horizon Prep in Rancho Santa Fe saw the biggest decline in personal belief exemptions between the 2015-16 and 2016-17 school year. It went from 13 of 40 students (32.5%) having a personal belief exemption to none.

And precisely how many schools now have no students with personal belief exemptions and how does that compare to the 2015-16 school year?

In [13]:
vax_sd[(vax_sd.pbe_1617 == 0.00)].shape

(177, 40)

And how many schools with suppressed data have at most 5 percent of students with personal belief exemptions?

In [14]:
vax_sd[(vax_sd.pbe_suppressed.str.contains('≤1%', na=False)) | 
       (vax_sd.pbe_suppressed.str.contains('≤2%', na=False)) | 
      (vax_sd.pbe_suppressed.str.contains('≤5%', na=False))].shape

(240, 40)

For comparison, how many schools in the 2015-16 school year had no students with personal belief exemptions?

In [15]:
vax_sd[(vax_sd.pbe_1516 == 0.00)].shape

(154, 40)

Which schools have the lowest proportion of their kindergarteners that are up-to-date on their vaccinations?

In [16]:
vax_sd.sort_values('uptodate_pct_1617', ascending=True).head()

school_code county_1617 pub_priv_1617            district_1617  \
265     0132886   SAN DIEGO        PUBLIC          BONSALL UNIFIED   
268     0127084   SAN DIEGO        PUBLIC  MOUNTAIN EMPIRE UNIFIED   
267     7052608   SAN DIEGO       PRIVATE                      NaN   
264     0132472   SAN DIEGO        PUBLIC           WARNER UNIFIED   
230     0132506   SAN DIEGO        PUBLIC        DEHESA ELEMENTARY   

          city_1617                                        school_1617  \
265         BONSALL                           PATHWAYS ACADEMY CHARTER   
268   THOUSAND OAKS  ACADEMY OF ARTS AND SCIENCES: DEL MAR MIDDLE &...   
267       SAN DIEGO             HERITAGE CHRISTIAN SCHOOL OF SAN DIEGO   
264  WARNER SPRINGS     CALIFORNIA PACIFIC CHARTER SCHOOLS - SAN DIEGO   
230        EL CAJON                     INSPIRE CHARTER SCHOOL - SOUTH   

     enrollment_1617  uptodate_1617  uptodate_pct_1617 uptodate_suppressed  \
265               40           3.00               7.50                 NaN   
268               22           4.00              18.18                 NaN   
267               30           7.00              23.33                 NaN   
264              260          92.00              35.38                 NaN   
230              525         210.00              40.00                 NaN   

     conditional_1617  conditional_pct_1617 conditional_suppressed  pme_1617  \
265             12.00                 30.00                    NaN      1.00   
268              0.00                  0.00                    NaN      0.00   
267              0.00                  0.00                    NaN      2.00   
264             26.00                 10.00                    NaN      4.00   
230              0.00                  0.00                    NaN      0.00   

     pme_pct_1617 pme_suppressed  pbe_1617  pbe_pct_1617 pbe_suppressed  \
265          2.50            NaN      8.00         20.00            NaN   
268          0.00            NaN      0.00          0.00            NaN   
267          6.67            NaN      2.00          6.67            NaN   
264          1.54            NaN     50.00         19.23            NaN   
230          0.00            NaN    315.00         60.00            NaN   

     others  others_pct others_suppressed county_1516 pub_priv_1516  \
265   16.00       40.00               NaN   SAN DIEGO        PUBLIC   
268   18.00       81.82               NaN         NaN           NaN   
267   19.00       63.33               NaN   SAN DIEGO       PRIVATE   
264   88.00       33.85               NaN   SAN DIEGO        PUBLIC   
230    0.00        0.00               NaN         NaN           NaN   

                district_1516       city_1516  \
265  BONSALL UNION ELEMENTARY         BONSALL   
268                       NaN             NaN   
267                       NaN       SAN DIEGO   
264            WARNER UNIFIED  WARNER SPRINGS   
230                       NaN             NaN   

                                school_1516  enrollment_1516  uptodate_1516  \
265                        PATHWAYS CHARTER           136.00          75.00   
268                                     NaN              nan            nan   
267  HERITAGE CHRISTIAN SCHOOL OF SAN DIEGO            27.00          21.00   
264  CALIFORNIA PACIFIC CHARTER SCHOOL - SD            78.00          25.00   
230                                     NaN              nan            nan   

     uptodate_pct_1516  conditional_1516  conditional_pct_1516  pme_1516  \
265              55.15             19.00                 13.97      0.00   
268                nan               nan                   nan       nan   
267              77.78              0.00                  0.00      0.00   
264              32.05              3.00                  3.85      0.00   
230                nan               nan                   nan       nan   

     pme_pct_1516  pbe_1516  pbe_pct_1516  \
265          0.00     42.00         30

So only 7.5% of the kindergarteners at Pathways Academy Charter in the Bonsall Unified school distrct are up-to-date on their vaccinations. Now what if we look at those schools where the exact number of students has been suppressed? Are any below 7.5%?

In [17]:
vax_sd.sort_values('uptodate_suppressed', ascending=True).head()

school_code county_1617 pub_priv_1617      district_1617    city_1617  \
447     6150312   SAN DIEGO       PRIVATE                NaN  CHULA VISTA   
508     6975973   SAN DIEGO       PRIVATE                NaN        VISTA   
474     6040117   SAN DIEGO        PUBLIC  SAN DIEGO UNIFIED    SAN DIEGO   
473     0123778   SAN DIEGO        PUBLIC  SAN DIEGO UNIFIED    SAN DIEGO   
472     6119598   SAN DIEGO        PUBLIC  SAN DIEGO UNIFIED    SAN DIEGO   

                           school_1617  enrollment_1617  uptodate_1617  \
447       MATER DEI JUAN DIEGO ACADEMY               30            nan   
508                    VISTA CHRISTIAN               23            nan   
474                   ROWAN ELEMENTARY               40            nan   
473       OLD TOWN ACADEMY K-8 CHARTER               28            nan   
472  KING-CHAVEZ ACADEMY OF EXCELLENCE               24            nan   

     uptodate_pct_1617 uptodate_suppressed  conditional_1617  \
447                nan                 ≤5%               nan   
508                nan                ≥95%               nan   
474                nan                ≥95%               nan   
473                nan                ≥95%               nan   
472                nan                ≥95%               nan   

     conditional_pct_1617 conditional_suppressed  pme_1617  pme_pct_1617  \
447                   nan                    ≤5%       nan           nan   
508                   nan                    ≤5%       nan           nan   
474                   nan                    ≤5%       nan           nan   
473                   nan                    ≤5%       nan           nan   
472                   nan                    ≤5%       nan           nan   

    pme_suppressed  pbe_1617  pbe_pct_1617 pbe_suppressed  others  others_pct  \
447            ≤5%       nan           nan            ≤5%     nan         nan   
508            ≤5%       nan           nan            ≤5%     nan         nan   
474            ≤5%       nan           nan            ≤5%     nan         nan   
473            ≤5%       nan           nan            ≤5%     nan         nan   
472            ≤5%       nan           nan            ≤5%     nan         nan   

    others_suppressed county_1516 pub_priv_1516      district_1516  city_1516  \
447               ≤5%         NaN           NaN                NaN        NaN   
508               ≤5%   SAN DIEGO       PRIVATE                NaN      VISTA   
474               ≤5%   SAN DIEGO        PUBLIC  SAN DIEGO UNIFIED  SAN DIEGO   
473               ≤5%   SAN DIEGO        PUBLIC  SAN DIEGO UNIFIED  SAN DIEGO   
472               ≤5%   SAN DIEGO        PUBLIC  SAN DIEGO UNIFIED  SAN DIEGO   

                           school_1516  enrollment_1516  uptodate_1516  \
447                                NaN              nan            nan   
508                    VISTA CHRISTIAN            25.00          20.00   
474                   ROWAN ELEMENTARY            45.00          44.00   
473       OLD TOWN ACADEMY K-8 CHARTER            28.00          16.00   
472  KING-CHAVEZ ACADEMY OF EXCELLENCE            24.00          24.00   

     uptodate_pct_1516  conditional_1516  conditional_pct_1516  pme_1516  \
447                nan               nan                   nan       nan   
508              80.00              3.00                 12.00      0.00   
474              97.78              1.00                  2.22      0.00   
473              57.14              0.00                  0.00      0.00   
472             100.00              0.00                  0.00      0.00   

     pme_pct_1516  pbe_1516  pbe_pct_1516                             school  \
447           nan       nan           nan                                NaN   
508          0.00      2.00          8.00                                NaN   
474          0.00      0.00          0.00                   Rowan Elementary   
473          0.00      2.00          7.14       Old Town A

Yes, there is one school with an up-to-date percentage less than or equal to five percent: Mater Dei Juan Diego Academy. The private school in Chula Vista has an enrollment of 30 students.

And which school has the highest proportion of students with personal belief exemptions?

In [18]:
vax_sd.sort_values('pbe_pct_1617', ascending=False).head()

school_code county_1617 pub_priv_1617              district_1617  \
230     0132506   SAN DIEGO        PUBLIC          DEHESA ELEMENTARY   
229     7102098   SAN DIEGO       PRIVATE                        NaN   
265     0132886   SAN DIEGO        PUBLIC            BONSALL UNIFIED   
264     0132472   SAN DIEGO        PUBLIC             WARNER UNIFIED   
259     6120893   SAN DIEGO        PUBLIC  SPENCER VALLEY ELEMENTARY   

          city_1617                                     school_1617  \
230        EL CAJON                  INSPIRE CHARTER SCHOOL - SOUTH   
229        CARLSBAD                       SANDERLING WALDORF SCHOOL   
265         BONSALL                        PATHWAYS ACADEMY CHARTER   
264  WARNER SPRINGS  CALIFORNIA PACIFIC CHARTER SCHOOLS - SAN DIEGO   
259     SIMI VALLEY          CALIFORNIA VIRTUAL ACADEMY @ SAN DIEGO   

     enrollment_1617  uptodate_1617  uptodate_pct_1617 uptodate_suppressed  \
230              525         210.00              40.00                 NaN   
229               21           9.00              42.86                 NaN   
265               40           3.00               7.50                 NaN   
264              260          92.00              35.38                 NaN   
259               73          49.00              67.12                 NaN   

     conditional_1617  conditional_pct_1617 conditional_suppressed  pme_1617  \
230              0.00                  0.00                    NaN      0.00   
229              1.00                  4.76                    NaN      5.00   
265             12.00                 30.00                    NaN      1.00   
264             26.00                 10.00                    NaN      4.00   
259              0.00                  0.00                    NaN      4.00   

     pme_pct_1617 pme_suppressed  pbe_1617  pbe_pct_1617 pbe_suppressed  \
230          0.00            NaN    315.00         60.00            NaN   
229         23.81            NaN      6.00         28.57            NaN   
265          2.50            NaN      8.00         20.00            NaN   
264          1.54            NaN     50.00         19.23            NaN   
259          5.48            NaN     14.00         19.18            NaN   

     others  others_pct others_suppressed county_1516 pub_priv_1516  \
230    0.00        0.00               NaN         NaN           NaN   
229    0.00        0.00               NaN   SAN DIEGO       PRIVATE   
265   16.00       40.00               NaN   SAN DIEGO        PUBLIC   
264   88.00       33.85               NaN   SAN DIEGO        PUBLIC   
259    6.00        8.22               NaN   SAN DIEGO        PUBLIC   

                 district_1516       city_1516  \
230                        NaN             NaN   
229                        NaN        CARLSBAD   
265   BONSALL UNION ELEMENTARY         BONSALL   
264             WARNER UNIFIED  WARNER SPRINGS   
259  SPENCER VALLEY ELEMENTARY     SIMI VALLEY   

                                school_1516  enrollment_1516  uptodate_1516  \
230                                     NaN              nan            nan   
229               SANDERLING WALDORF SCHOOL            32.00           9.00   
265                        PATHWAYS CHARTER           136.00          75.00   
264  CALIFORNIA PACIFIC CHARTER SCHOOL - SD            78.00          25.00   
259    CALIFORNIA VIRTUAL ACADEMY SAN DIEGO           108.00          90.00   

     uptodate_pct_1516  conditional_1516  conditional_pct_1516  pme_1516  \
230                nan               nan                   nan       nan   
229              28.13              0.00                  0.00      0.00   
265              55.15             19.00                 13.97      0.00   
264              32.05              3.00                  3.85      0.00   
259              83.33              0.00                  0.00      0.00   

     pme_pct_1516  pbe_1516  pbe_pct_1516  \
230           nan       nan           

How many kindergartens saw an increase in the proportion of their students that are up-to-date on their vaccinations?

In [19]:
vax_sd[vax_sd.pct_change_uptodate > 0].shape

(141, 40)

And how many kindergartens saw a decrease?

In [20]:
vax_sd[vax_sd.pct_change_uptodate < 0].shape

(121, 40)

In [21]:
vax_sd.to_excel('vax_sd.xlsx')

When we analyzed the data two years ago, we found that wealthier schools (as measured by public vs. private and the proportion of students receiving free-and-reduced-price lunch) had lower personal belief exemption rates than poorer schools. Is that still the case?  
  
To find out, let's create two new dataframes, one of the public kindergartens and one of the private kindergartens, and then calcuate the personal belief exemption rates for each set of data. We'll also need to exclude those kindergartens where data has been partially masked.

In [22]:
vax_sd_pub = vax_sd[(vax_sd.pub_priv_1617 == 'PUBLIC') & (vax_sd.pbe_1617.notnull())]
vax_sd_priv = vax_sd[(vax_sd.pub_priv_1617 == 'PRIVATE') & (vax_sd.pbe_1617.notnull())] 

In [23]:
# Calculate the PBE exemption rate for public schools
vax_sd_pub.pbe_1617.sum() / vax_sd_pub.enrollment_1617.sum()

0.02385507836352534

In [24]:
# Calculate the PBE exemption rate for private schools
vax_sd_priv.pbe_1617.sum() / vax_sd_priv.enrollment_1617.sum()

0.025763358778625955

So the rates for public and private schools are similar. What about when we look at the rates for the proportion of students receiving free-and-reduced-price lunch?  
  
The first step is to create two new dataframes, one of the public kindergartens where more than half of the student population qualified for free or reduced-cost meals and one of the public kindergartens where half or less of the student population qualified for free or reduced-cost meals.

In [25]:
vax_sd_high_frpm = vax_sd_pub[vax_sd_pub.frpm_pct > 0.5]
vax_sd_low_frpm = vax_sd_pub[vax_sd_pub.frpm_pct <= 0.5]

In [26]:
# Calculate the PBE exemption rate for high FRPM kindergartens
vax_sd_high_frpm.pbe_1617.sum() / vax_sd_high_frpm.enrollment_1617.sum()

0.004946219674962707

In [27]:
# Calculate the PBE exemption rate for low FRPM kindergartens
vax_sd_low_frpm.pbe_1617.sum() / vax_sd_low_frpm.enrollment_1617.sum()

0.04421711193777477

## Import school address data

Now that we've analyzed the latest year of data, we need to prepare it for mapping. Let's start by bring in the Excel spreadsheets of school addresses, starting with the public schools file.

In [28]:
# Read the CDS code, zip code, latitude and longitude columns as strings, specify the columns to import and assign new column names
pubaddress = pd.read_excel('data/pubschools.xlsx', converters={'CDSCode':str, 'Zip':str, 'Latitude':str,'Longitude':str}, parse_cols='A,G:H,J:L,AL:AM',
                           names=('cds_code','school','street','city','zip','state','latitude','longitude'))
pubaddress.info() # 17,707 records returned, as expected

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17707 entries, 0 to 17706
Data columns (total 8 columns):
cds_code     17707 non-null object
school       16337 non-null object
street       17415 non-null object
city         17415 non-null object
zip          17415 non-null object
state        17415 non-null object
latitude     12899 non-null object
longitude    12899 non-null object
dtypes: object(8)
memory usage: 1.1+ MB


In [29]:
pubaddress.head(1)

cds_code school                  street     city         zip state  \
0  01100170000000    NaN  313 West Winton Avenue  Hayward  94544-1136    CA   

    latitude   longitude  
0  37.658212  -122.09713

Now we need to slice out the last seven digits of the CDS code to get the school code.

In [30]:
pubaddress['school_code'] = pubaddress['cds_code'].str[-7:]
pubaddress.drop('cds_code', axis=1, inplace=True) # Delete the CDS code column as it is no longer needed
pubaddress.head()

school  \
0                                            NaN   
1                            FAME Public Charter   
2         Envision Academy for Arts & Technology   
3  Aspire California College Preparatory Academy   
4        Community School for Creative Education   

                             street      city         zip state   latitude  \
0            313 West Winton Avenue   Hayward  94544-1136    CA  37.658212   
1  39899 Balentine Drive, Suite 335    Newark  94560-5359    CA  37.521436   
2               1515 Webster Street   Oakland  94612-3355    CA  37.804520   
3             2125 Jefferson Avenue  Berkeley  94703-1414    CA  37.868991   
4      2111 International Boulevard   Oakland  94606-4903    CA  37.784648   

    longitude school_code  
0  -122.09713     0000000  
1  -121.99391     0109835  
2  -122.26815     0112607  
3  -122.27844     0118489  
4  -122.23863     0123968

And finally, let's merge the street, city, state and zip columns into a single address column on which we will eventually map the data.

In [31]:
pubaddress['address'] = pubaddress['street'] + ' ' + pubaddress['city'] + ', ' + pubaddress['state'] + ' ' + pubaddress['zip']
pubaddress.drop(['street', 'city',  'state', 'zip'], axis=1, inplace=True) # Delete the street, city, state and zip columns as they are no longer needed
pubaddress.head()

school   latitude   longitude  \
0                                            NaN  37.658212  -122.09713   
1                            FAME Public Charter  37.521436  -121.99391   
2         Envision Academy for Arts & Technology  37.804520  -122.26815   
3  Aspire California College Preparatory Academy  37.868991  -122.27844   
4        Community School for Creative Education  37.784648  -122.23863   

  school_code                                            address  
0     0000000      313 West Winton Avenue Hayward, CA 94544-1136  
1     0109835  39899 Balentine Drive, Suite 335 Newark, CA 94...  
2     0112607         1515 Webster Street Oakland, CA 94612-3355  
3     0118489      2125 Jefferson Avenue Berkeley, CA 94703-1414  
4     0123968  2111 International Boulevard Oakland, CA 94606...

And now let's import the addresses for the private schools.

In [32]:
# Read the CDS code and zip code columns as strings, specify the columns to import and assign new column names
privaddress = pd.read_excel('data/privschools1617.xlsx', converters={'CDS Code': str, 'Zip': str}, 
                            parse_cols='C:H', names=('cds_code', 'school', 'street', 'city', 'state', 'zip'))
privaddress.info() # 2,939 records returned, as expected

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2939 entries, 0 to 2938
Data columns (total 6 columns):
cds_code    2772 non-null object
school      2939 non-null object
street      2939 non-null object
city        2939 non-null object
state       2939 non-null object
zip         2939 non-null object
dtypes: object(6)
memory usage: 137.8+ KB


In [33]:
privaddress.head(1)

cds_code                             school  \
0  01612596910343  Agnes Memorial Christian Academey   

                         street     city state    zip  
0  2372 International Boulevard  Oakland    CA  94601

As with the public schools file, we need to slice out the last seven digits of the CDS code to get the school code.

In [34]:
privaddress['school_code'] = privaddress['cds_code'].str[-7:]
privaddress.drop('cds_code', axis=1, inplace=True) # Delete the CDS code column as it is no longer needed
privaddress.head()

school  \
0                  Agnes Memorial Christian Academey   
1                         All Saints Catholic School   
2                      Alsion Montessori Middle/High   
3           American International Montessori School   
4  Anchor Education, Inc. dba Sterling West -- Sa...   

                            street           city state    zip school_code  
0     2372 International Boulevard        Oakland    CA  94601     6910343  
1              22870 Second Street        Hayward    CA  94541     6965909  
2                750 Witherly Lane        Fremont    CA  94539     7103799  
3  3339 Martin Luther King Jr. Way       Berkeley    CA  94703     6135743  
4           19234 Lake Chabot Road  Castro Valley    CA  94577     6133441

And finally, let's merge the street, city, state and zip columns into a single address column on which we will eventually map the data.

In [35]:
privaddress['address'] = privaddress['street'] + ' ' + privaddress['city'] + ', ' + privaddress['state'] + ' ' + privaddress['zip']
privaddress.drop(['street', 'city',  'state', 'zip'], axis=1, inplace=True) # Delete the street, city, state and zip columns as they are no longer needed
privaddress.head()

school school_code  \
0                  Agnes Memorial Christian Academey     6910343   
1                         All Saints Catholic School     6965909   
2                      Alsion Montessori Middle/High     7103799   
3           American International Montessori School     6135743   
4  Anchor Education, Inc. dba Sterling West -- Sa...     6133441   

                                             address  
0     2372 International Boulevard Oakland, CA 94601  
1              22870 Second Street Hayward, CA 94541  
2                750 Witherly Lane Fremont, CA 94539  
3  3339 Martin Luther King Jr. Way Berkeley, CA 9...  
4     19234 Lake Chabot Road Castro Valley, CA 94577

### Concatenate, join and export the data

Now we need to concatenate (combine) the public and private school address dataframes so we can then join them with the 2016-17 vaccinations/lunch dataframe.

In [36]:
all_addresses = pd.concat([pubaddress, privaddress])
all_addresses.info() # 20,646 records returned, as expected

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20646 entries, 0 to 2938
Data columns (total 5 columns):
address        20354 non-null object
latitude       12899 non-null object
longitude      12899 non-null object
school         19276 non-null object
school_code    20479 non-null object
dtypes: object(5)
memory usage: 967.8+ KB


Merge the addresses and vaccinations/school lunch dataframes.

In [37]:
vax_sd_addresses = pd.merge(vax_sd, all_addresses, on='school_code', how='left', suffixes=('_vax', '_addresses'))
vax_sd_addresses.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 511 entries, 0 to 510
Data columns (total 44 columns):
school_code               511 non-null object
county_1617               511 non-null object
pub_priv_1617             511 non-null object
district_1617             452 non-null object
city_1617                 511 non-null object
school_1617               511 non-null object
enrollment_1617           511 non-null int64
uptodate_1617             269 non-null float64
uptodate_pct_1617         269 non-null float64
uptodate_suppressed       242 non-null object
conditional_1617          269 non-null float64
conditional_pct_1617      269 non-null float64
conditional_suppressed    242 non-null object
pme_1617                  269 non-null float64
pme_pct_1617              269 non-null float64
pme_suppressed            242 non-null object
pbe_1617                  269 non-null float64
pbe_pct_1617              269 non-null float64
pbe_suppressed            242 non-null object
others        

And finally, export the data as an Excel file.

In [38]:
vax_sd_addresses.to_excel('data/vax_sd_addresses.xlsx')